docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,546 @@
+}
+
+// Spec defines the desired state of NetworkQoS
+type Spec struct {
+// netAttachRefs points to a list of objects which could be either NAD, UDN, or Cluster UDN.
+// In the case of NAD, the network type could be of type Layer-3, Layer-2, or Localnet.
+// If not specified, then the primary network of the selected Pods will be chosen.
+// +optional
+// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="netAttachRefs is immutable"
+NetworkAttachmentRefs []corev1.ObjectReference `json:"netAttachRefs,omitempty"`
+
+// podSelector applies the NetworkQoS rule only to the pods in the namespace whose label
+// matches this definition. This field is optional, and in case it is not set
+// results in the rule being applied to all pods in the namespace.
+// +optional
+PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+
+// egress a collection of Egress NetworkQoS rule objects
+Egress []Rule `json:"egress"`
+}
+
+type Rule struct {
+// priority The NetworkQoS rule’s priority. Rules with numerically higher
+// priority takes precedence over those with lower. The usable range is
+// from 10001 to 10100. Using the priority value outside of this range will
+// result in an error.	
+// +kubebuilder:validation:Maximum:=10100
+// +kubebuilder:validation:Minimum:=10001
+Priority int `json:"priority"`
+
+// dscp marking value for matching pods' traffic.
+// +kubebuilder:validation:Maximum:=63
+// +kubebuilder:validation:Minimum:=0
+DSCP int `json:"dscp"`
+
+// classifier The classifier on which packets should match
+// to apply the NetworkQoS Rule.
+// This field is optional, and in case it is not set the rule is applied
+// to all egress traffic regardless of the destination.
+// +optional
+Classifier Classifier `json:"classifier"`
+
+// +optional
+Bandwidth Bandwidth `json:"bandwidth"`
+}
+
+type Classifier struct {
+// +optional
+To []Destination `json:"to"`
+
+// +optional
+Port Port `json:"port"`
+}
+
+// Bandwidth controls the maximum of rate traffic that can be sent
+// or received on the matching packets.
+type Bandwidth struct {
+// rate The value of rate limit in kbps. Traffic over the limit
+// will be dropped.
+// +kubebuilder:validation:Minimum:=1
+// +kubebuilder:validation:Maximum:=4294967295
+// +optional
+Rate uint32 `json:"rate"`
+
+// burst The value of burst rate limit in kilobits.
+// This also needs rate to be specified.
+// +kubebuilder:validation:Minimum:=1
+// +kubebuilder:validation:Maximum:=4294967295
+// +optional
+Burst uint32 `json:"burst"`
+}
+
+// Port specifies destination protocol and port on which NetworkQoS
+// rule is applied
+type Port struct {
+// protocol (tcp, udp, sctp) that the traffic must match.
+// +kubebuilder:validation:Pattern=^TCP|UDP|SCTP$
+// +optional
+Protocol string `json:"protocol"`
+
+// port that the traffic must match
+// +kubebuilder:validation:Minimum:=1
+// +kubebuilder:validation:Maximum:=65535
+// +optional
+Port int32 `json:"port"`
```

Comment:
do we want an endPort here?

Reply:
We don't see the need for it now. If it becomes urgent, then we can add the endPort later like how we did for the K8s NetworkPolicy.